doobie/src/main/scala/tamer/db/package.scala

```diff
@@ -0,0 +1,111 @@
+package tamer
+
+import cats.effect.Blocker
+import com.sksamuel.avro4s.{Decoder, Encoder, SchemaFor}
+import doobie.Query0
+import doobie.hikari.HikariTransactor
+import doobie.implicits.{toDoobieStreamOps, _}
+import doobie.util.transactor.Transactor
+import eu.timepit.refined.auto._
+import fs2.Stream
+import log.effect.LogWriter
+import log.effect.zio.ZioLogWriter.log4sFromName
+import tamer.db.Compat.toIterable
+import tamer.db.ConfigDb.{DbConfig, QueryConfig}
+import tamer.db.Db.{Datable, TimeSegment, _}
+import tamer.kafka.Kafka
+import zio.blocking.Blocking
+import zio.interop.catz._
+import zio.{Queue, Task, UIO, ZIO, _}
+
+import java.sql.SQLException
+import java.time.{Duration, Instant}
+import scala.concurrent.ExecutionContext
+
+package object db {
+  type DbTransactor  = Has[Transactor[Task]]
+  type TamerDBConfig = DbTransactor with QueryConfig
+
+  private[this] implicit final class InstantOps(private val instant: Instant) extends AnyVal {
+    def orNow: UIO[Instant] =
+      UIO(Instant.now).map {
+        case now if instant.isAfter(now) => now
+        case _                           => instant
+      }
+  }
+
+  final def mkSetup[K <: Product: Encoder: Decoder: SchemaFor, V <: Product with Datable: Ordering: Encoder: Decoder: SchemaFor](
+      queryBuilder: TimeSegment => Query0[V]
+  )(earliest: Instant, tumblingStep: Duration, keyExtract: V => K): Setup[K, V, TimeSegment] = {
+
+    val timeSegment = TimeSegment(earliest, earliest.plus(tumblingStep))
+
+    val qBuilder = new QueryBuilder[V, TimeSegment] {
+      override val queryId: Int                         = queryBuilder(timeSegment).sql.hashCode
+      override def query(state: TimeSegment): Query0[V] = queryBuilder(state)
+    }
+
+    def stateFold(timeSegment: TimeSegment)(queryResult: QueryResult[V]): UIO[TimeSegment] =
+      if (queryResult.results.isEmpty) timeSegment.to.plus(tumblingStep).orNow.map(TimeSegment(timeSegment.from, _))
+      else {
+        val mostRecent = queryResult.results.max.instant
+        mostRecent.plus(tumblingStep).orNow.map(TimeSegment(mostRecent, _))
+      }
+
+    Setup[K, V, TimeSegment](qBuilder, timeSegment, keyExtract, stateFold)
+  }
+
+  private[this] val logTask: Task[LogWriter[Task]] = log4sFromName.provide("tamer.db")
+
+  final def iteration[K <: Product, V <: Product](
+      setup: Setup[K, V, TimeSegment]
+  )(state: TimeSegment, q: Queue[(K, V)]): ZIO[TamerDBConfig, TamerError, TimeSegment] =
+    (for {
+      log   <- logTask
+      cfg   <- ConfigDb.queryConfig
+      tnx   <- ZIO.service[Transactor[Task]]
+      query <- UIO(setup.queryBuilder.query(state))
+      _     <- log.info(s"running ${query.sql} with params derived from $state") // TODO: make this debug
+      start <- UIO(Instant.now())
```

Comment:
Time to fix these, we should not be measuring these durations via `Instant.now()` but `System.nanoTime()`

Reply:
Ok, but I have no idea why

Reply:
oh you mean `zio.clock.nanoTime/instant`?

Reply:
Yup - `zio.clock.nanoTime`

The reason is nanoTime gives you a monotonic value (representing the number of nanoseconds since system boot) while instant gives you a non-monotonic value (millis since epoch). The latter is subject to changes made to the system clock (including it going backwards), usually as result of NTP adjustments. See also https://medium.com/@kralka/its-about-time-d42e68dbe1df

Reply:
Let's track this separately I fell this PR might cover too many concerns